torch/csrc/api/include/torch/data/detail/data-shuttle.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <torch/data/detail/queue.h>
+
+#include <ATen/Error.h>
+#include <ATen/optional.h>
+
+#include <chrono>
+#include <queue>
+#include <utility>
+
+namespace torch {
+namespace data {
+namespace detail {
+
+template <typename Job, typename Result>
+class DataShuttle {
+ public:
+  using JobType = Job;
+  using ResultType = Result;
+
+  enum class Event { kPush, kExhausted };
+
+  void push_job(Job&& job) {
+    jobs_.push(std::move(job));
+    log_.push(Event::kPush);
+  }
+
+  void push_result(Result&& result) {
+    results_.push(std::move(result));
+  }
+
+  Job pop_job() {
+    return jobs_.pop();
+  }
+
+  at::optional<Result> pop_result(
+      at::optional<std::chrono::seconds> timeout = at::nullopt) {
+    const auto event = log_.front();
+    log_.pop();
+    switch (event) {
+      case Event::kPush:
+        return results_.pop(timeout);
+      case Event::kExhausted:
+        return at::nullopt;
+      default:
+        AT_ERROR("Unhandled event");
+    }
+  }
+
+  void exhausted() {
+    log_.push(Event::kExhausted);
```

Comment:
It could be replaced with a counter and boolean, that's true. I found this "log" idea quite clear conceptually. I'll try your suggestion and see if it makes the code clearer.

Reply:
IIRC we use the counter in Python